File: go/tests/aethernet_mesh_capability_test.go
// aethernet_mesh_capability_test.go
//
// Verifies CircleAI.AetherNet.MeshCapabilityRegistry port
// (aethernet_mesh_capability.go): Upsert replaces per-peer, Remove is
// idempotent, List honours the staleness filter, Find matches model
// (case-insensitive) + min-KV and sorts by spare budget descending, and the
// broadcasters (Null no-op + loopback that feeds the registry).

package circleai_test

import (
	"context"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func mkAd(peer, model string, freeKv int, at time.Time) circleai.MeshCapabilityAdvertisement {
	return circleai.MeshCapabilityAdvertisement{
		PeerID:              peer,
		ModelID:             model,
		FreeKvTokens:        freeKv,
		Tier:                circleai.DeviceTierPhone,
		ContextWindowTokens: 2048,
		AdvertisedAtUTC:     at,
	}
}

func TestMeshRegistry_UpsertReplacesPerPeer(t *testing.T) {
	reg := circleai.NewInMemoryMeshCapabilityRegistry()
	ctx := context.Background()
	now := time.Now().UTC()

	_ = reg.Upsert(ctx, mkAd("p1", "Qwen3-1.7B-MNN", 1000, now))
	_ = reg.Upsert(ctx, mkAd("p1", "Qwen3-1.7B-MNN", 2000, now)) // replace, not add

	all := reg.List(nil)
	if len(all) != 1 {
		t.Fatalf("upsert should replace: got %d entries want 1", len(all))
	}
	if all[0].FreeKvTokens != 2000 {
		t.Errorf("latest advertisement should win: got %d", all[0].FreeKvTokens)
	}
}

func TestMeshRegistry_RemoveIdempotent(t *testing.T) {
	reg := circleai.NewInMemoryMeshCapabilityRegistry()
	ctx := context.Background()
	_ = reg.Upsert(ctx, mkAd("p1", "m", 10, time.Now().UTC()))

	removed, _ := reg.Remove(ctx, "p1")
	if !removed {
		t.Error("first remove should report true")
	}
	removed, _ = reg.Remove(ctx, "p1")
	if removed {
		t.Error("second remove should report false (idempotent)")
	}
	if len(reg.List(nil)) != 0 {
		t.Error("registry should be empty after remove")
	}
}

func TestMeshRegistry_ListStalenessFilter(t *testing.T) {
	fixed := time.Date(2026, 7, 10, 12, 0, 0, 0, time.UTC)
	reg := circleai.NewInMemoryMeshCapabilityRegistry().WithClock(func() time.Time { return fixed })
	ctx := context.Background()

	_ = reg.Upsert(ctx, mkAd("fresh", "m", 10, fixed.Add(-10*time.Second)))
	_ = reg.Upsert(ctx, mkAd("stale", "m", 10, fixed.Add(-90*time.Second)))

	// No filter → both.
	if got := len(reg.List(nil)); got != 2 {
		t.Errorf("unfiltered list got %d want 2", got)
	}
	// 60s window → only the fresh one.
	window := 60 * time.Second
	got := reg.List(&window)
	if len(got) != 1 || got[0].PeerID != "fresh" {
		t.Errorf("staleness filter got %+v, want only 'fresh'", got)
	}
}

func TestMeshRegistry_FindMatchesAndSorts(t *testing.T) {
	fixed := time.Date(2026, 7, 10, 12, 0, 0, 0, time.UTC)
	reg := circleai.NewInMemoryMeshCapabilityRegistry().WithClock(func() time.Time { return fixed })
	ctx := context.Background()

	_ = reg.Upsert(ctx, mkAd("low", "Qwen3-1.7B-MNN", 500, fixed))
	_ = reg.Upsert(ctx, mkAd("high", "qwen3-1.7b-mnn", 4000, fixed)) // case-insensitive match
	_ = reg.Upsert(ctx, mkAd("mid", "Qwen3-1.7B-MNN", 2000, fixed))
	_ = reg.Upsert(ctx, mkAd("other", "Gemma-2B", 9000, fixed)) // different model

	// min 1000 KV → excludes "low" (500) and "other" (wrong model).
	got := reg.Find("Qwen3-1.7B-MNN", 1000, nil)
	if len(got) != 2 {
		t.Fatalf("Find got %d want 2 (%+v)", len(got), got)
	}
	// Sorted by spare budget DESCENDING → high (4000) before mid (2000).
	if got[0].PeerID != "high" || got[1].PeerID != "mid" {
		t.Errorf("Find ordering wrong: got %s,%s want high,mid", got[0].PeerID, got[1].PeerID)
	}

	// min 0 includes low too (still excludes wrong model).
	got = reg.Find("Qwen3-1.7B-MNN", 0, nil)
	if len(got) != 3 {
		t.Errorf("Find min0 got %d want 3", len(got))
	}
	if got[0].PeerID != "high" || got[2].PeerID != "low" {
		t.Errorf("Find min0 ordering: got %s..%s", got[0].PeerID, got[2].PeerID)
	}
}

func TestMeshRegistry_FindStalenessFilter(t *testing.T) {
	fixed := time.Date(2026, 7, 10, 12, 0, 0, 0, time.UTC)
	reg := circleai.NewInMemoryMeshCapabilityRegistry().WithClock(func() time.Time { return fixed })
	ctx := context.Background()

	_ = reg.Upsert(ctx, mkAd("fresh", "m", 1000, fixed.Add(-5*time.Second)))
	_ = reg.Upsert(ctx, mkAd("stale", "m", 5000, fixed.Add(-120*time.Second)))

	window := 60 * time.Second
	got := reg.Find("m", 0, &window)
	if len(got) != 1 || got[0].PeerID != "fresh" {
		t.Errorf("Find staleness got %+v want only 'fresh'", got)
	}
}

func TestMeshRegistry_UpsertBlankPeerPanics(t *testing.T) {
	reg := circleai.NewInMemoryMeshCapabilityRegistry()
	defer func() {
		if recover() == nil {
			t.Error("expected panic on blank PeerID")
		}
	}()
	_ = reg.Upsert(context.Background(), mkAd("   ", "m", 1, time.Now().UTC()))
}

func TestNullMeshCapabilityBroadcaster_NoOp(t *testing.T) {
	err := circleai.NullMeshCapabilityBroadcasterInstance.Broadcast(context.Background(),
		mkAd("p1", "m", 1, time.Now().UTC()))
	if err != nil {
		t.Errorf("null broadcaster should never error: %v", err)
	}
}

func TestLoopbackBroadcaster_FeedsRegistry(t *testing.T) {
	reg := circleai.NewInMemoryMeshCapabilityRegistry()
	bc := circleai.NewLocalLoopbackMeshCapabilityBroadcaster(reg)
	ctx := context.Background()

	ad := mkAd("me", "Qwen3-1.7B-MNN", 3000, time.Now().UTC())
	if err := bc.Broadcast(ctx, ad); err != nil {
		t.Fatalf("broadcast: %v", err)
	}
	got := reg.Find("Qwen3-1.7B-MNN", 0, nil)
	if len(got) != 1 || got[0].PeerID != "me" {
		t.Errorf("loopback broadcast should land in registry, got %+v", got)
	}
}

File: typescript/tests/aethernet_mesh_capability.test.ts
// aethernet_mesh_capability.test.ts
//
// Verifies the CircleAI.AetherNet mesh capability discovery port (RT-12 v1):
//   InMemoryMeshCapabilityRegistry — upsert/replace, remove (idempotent),
//     list (+ staleAfter filter), find (case-insensitive model match, minFreeKv,
//     staleAfter, sorted by free budget descending)
//   NullMeshCapabilityBroadcaster — no-op
//   RegistryBackedMeshCapabilityBroadcaster — loopback into a registry

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { DeviceTier } from "../src/device/index";
import {
  InMemoryMeshCapabilityRegistry,
  NullMeshCapabilityBroadcaster,
  RegistryBackedMeshCapabilityBroadcaster,
  meshCapabilityAdvertisement,
  type MeshCapabilityAdvertisement,
} from "../src/aethernet/index";

function ad(
  peerId: string,
  modelId: string,
  freeKv: number,
  at: Date,
  tier: DeviceTier = DeviceTier.Phone,
): MeshCapabilityAdvertisement {
  return meshCapabilityAdvertisement(peerId, modelId, freeKv, tier, 2048, at);
}

describe("InMemoryMeshCapabilityRegistry", () => {
  it("upsert stores and replaces per peer", async () => {
    const reg = new InMemoryMeshCapabilityRegistry();
    const t = new Date();
    await reg.upsertAsync(ad("p1", "Qwen3-1.7B-MNN", 1000, t));
    assert.equal(reg.count, 1);
    // Replace the same peer's advertisement.
    await reg.upsertAsync(ad("p1", "Qwen3-1.7B-MNN", 2000, t));
    assert.equal(reg.count, 1);
    assert.equal(reg.list()[0].freeKvTokens, 2000);
  });

  it("rejects a blank peer id", async () => {
    const reg = new InMemoryMeshCapabilityRegistry();
    // Validation throws synchronously (mirrors C# ArgumentException.ThrowIfNullOrWhiteSpace
    // before the ValueTask is produced); wrap in an async fn so the sync throw
    // becomes a rejection for assert.rejects.
    await assert.rejects(async () => reg.upsertAsync(ad("   ", "M", 1, new Date())));
  });

  it("remove is idempotent and reports whether a peer was removed", async () => {
    const reg = new InMemoryMeshCapabilityRegistry();
    await reg.upsertAsync(ad("p1", "M", 1, new Date()));
    assert.equal(await reg.removeAsync("p1"), true);
    assert.equal(await reg.removeAsync("p1"), false);
    assert.equal(reg.count, 0);
  });

  it("list without staleAfter returns everything; with staleAfter filters old entries", async () => {
    const now = new Date("2026-07-10T12:00:00Z");
    const reg = new InMemoryMeshCapabilityRegistry();
    reg.nowUtc = () => now;

    const fresh = new Date(now.getTime() - 10_000); // 10s ago
    const stale = new Date(now.getTime() - 120_000); // 2min ago
    await reg.upsertAsync(ad("fresh", "M", 1, fresh));
    await reg.upsertAsync(ad("stale", "M", 1, stale));

    assert.equal(reg.list().length, 2);
    // 60s staleness window drops the 2-min-old entry.
    const recent = reg.list(60_000);
    assert.equal(recent.length, 1);
    assert.equal(recent[0].peerId, "fresh");
  });

  it("find matches modelId case-insensitively, respects minFreeKv, sorts by free budget desc", async () => {
    const reg = new InMemoryMeshCapabilityRegistry();
    const t = new Date();
    await reg.upsertAsync(ad("p1", "Qwen3-1.7B-MNN", 500, t));
    await reg.upsertAsync(ad("p2", "qwen3-1.7b-mnn", 3000, t)); // different case, same model
    await reg.upsertAsync(ad("p3", "Qwen3-1.7B-MNN", 1500, t));
    await reg.upsertAsync(ad("p4", "SomethingElse", 9000, t)); // different model

    const hits = reg.find("QWEN3-1.7B-MNN", 1000);
    assert.deepEqual(hits.map((h) => h.peerId), ["p2", "p3"]); // p1 filtered (500 < 1000); sorted desc
    assert.equal(hits[0].freeKvTokens, 3000);
  });

  it("find with staleAfter filters stale peers", async () => {
    const now = new Date("2026-07-10T12:00:00Z");
    const reg = new InMemoryMeshCapabilityRegistry();
    reg.nowUtc = () => now;
    await reg.upsertAsync(ad("fresh", "M", 1000, new Date(now.getTime() - 5_000)));
    await reg.upsertAsync(ad("stale", "M", 5000, new Date(now.getTime() - 300_000)));

    const hits = reg.find("M", 0, 60_000);
    assert.deepEqual(hits.map((h) => h.peerId), ["fresh"]);
  });

  it("find rejects a blank modelId", () => {
    const reg = new InMemoryMeshCapabilityRegistry();
    assert.throws(() => reg.find("  "));
  });
});

describe("NullMeshCapabilityBroadcaster", () => {
  it("broadcast is a no-op that resolves", async () => {
    await NullMeshCapabilityBroadcaster.instance.broadcastAsync(ad("p1", "M", 1, new Date()));
    // Nothing to assert beyond "did not throw / resolved".
    assert.ok(NullMeshCapabilityBroadcaster.instance);
  });
});

describe("RegistryBackedMeshCapabilityBroadcaster", () => {
  it("loops the advertisement into the bound registry", async () => {
    const reg = new InMemoryMeshCapabilityRegistry();
    const caster = new RegistryBackedMeshCapabilityBroadcaster(reg);
    await caster.broadcastAsync(ad("me", "Qwen3-1.7B-MNN", 2048, new Date(), DeviceTier.Workstation));
    assert.equal(caster.broadcastCount, 1);
    const found = reg.find("Qwen3-1.7B-MNN");
    assert.equal(found.length, 1);
    assert.equal(found[0].peerId, "me");
    assert.equal(found[0].tier, DeviceTier.Workstation);
  });
});
